diff -urp nginx-1.17.8/auto/cc/clang nginx-1.17.8-patched/auto/cc/clang
--- nginx-1.17.8/auto/cc/clang	2014-03-04 03:39:24.000000000 -0800
+++ nginx-1.17.8-patched/auto/cc/clang	2014-03-13 20:54:26.241413360 -0700
@@ -89,7 +89,7 @@ CFLAGS="$CFLAGS -Wconditional-uninitiali
 CFLAGS="$CFLAGS -Wno-unused-parameter"
 
 # stop on warning
-CFLAGS="$CFLAGS -Werror"
+#CFLAGS="$CFLAGS -Werror"
 
 # debug
 CFLAGS="$CFLAGS -g"
diff -urp nginx-1.17.8/auto/cc/gcc nginx-1.17.8-patched/auto/cc/gcc
--- nginx-1.17.8/auto/cc/gcc	2014-03-04 03:39:24.000000000 -0800
+++ nginx-1.17.8-patched/auto/cc/gcc	2014-03-13 20:54:13.301355329 -0700
@@ -168,7 +168,7 @@ esac
 
 
 # stop on warning
-CFLAGS="$CFLAGS -Werror"
+#CFLAGS="$CFLAGS -Werror"
 
 # debug
 CFLAGS="$CFLAGS -g"
diff -urp nginx-1.17.8/auto/cc/icc nginx-1.17.8-patched/auto/cc/icc
--- nginx-1.17.8/auto/cc/icc	2014-03-04 03:39:24.000000000 -0800
+++ nginx-1.17.8-patched/auto/cc/icc	2014-03-13 20:54:13.301355329 -0700
@@ -115,7 +115,7 @@ case "$NGX_ICC_VER" in
 esac
 
 # stop on warning
-CFLAGS="$CFLAGS -Werror"
+#CFLAGS="$CFLAGS -Werror"
 
 # debug
 CFLAGS="$CFLAGS -g"
